### PR TITLE
Treap 

### PR DESCRIPTION
**Pull request by dutinmeow**
I will format the problems later

### CODE REVIEW

File: content/6_Advanced/Treaps.mdx [hunk 1]
```diff
@@ -207,7 +254,252 @@ int main() {
...
+
+$$\texttt{index}(u) = |\{1 + \texttt{size}(P.\texttt{left}) \mid P \text{ is on path from } u \text{ to root}\}|$$
+
+Note that in an implicit treap, the merge function is largely unchanged because
+it does not depend on the key. In the split operation we go from the root down,
+so we simply maintain a running count of the size left subtree.
+
+An implementation of the split operation may look something like:
+
+```cpp
```
[the remark this answers is not in the document]
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:294]: done

**Review by dongliu0426 — APPROVED** [review]

**Comment by Aadit-Ambadkar** [comment]
> at least your code is consistently bad

kinda rude

File: content/6_Advanced/Treaps.mdx [hunk 2]
```diff
@@ -147,50 +178,66 @@ ll ans, inv;
...
-	x->sz = 1+getsz(x->c[0])+getsz(x->c[1]);
+	x->sz = 1 + getsz(x->c[0]) + getsz(x->c[1]);
 	return x;
 }
-pair<pt,pt> split(pt t, int v) { // >= v goes to the right
-	if (!t) return {t,t};
+	
+pair<pt, pt> split(pt t, int v) { // >= v goes to the right
+	if (!t)
+		return {t,t};
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:201]:
```suggestion
	if (!t) return {t,t};
```

File: content/6_Advanced/Treaps.mdx [hunk 3]
```diff
@@ -147,50 +178,66 @@ ll ans, inv;
...
+		auto p = split(t->c[1], v);
+		t->c[1] = p.f;
+		return {calc(t), p.s};
 	}
 }
 
 pt merge(pt a, pt b) {
-	if (!a || !b) return a?:b;
+	if (!a || !b)
+		return a?:b;
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:215]:
```suggestion
	if (!a || !b) return a?:b;
```

File: content/6_Advanced/Treaps.mdx [hunk 4]
```diff
@@ -147,50 +178,66 @@ ll ans, inv;
...
 pt go() {
-	int x; re(x);
-	if (x) return new tnode(x);
-	pt L = go(); pt R = go();
-	ll tot = (ll)L->sz*R->sz; inv = 0;
-	pt cur = merge(L,R); assert(inv <= tot); // dbg("HUH",tot,inv);
+	int x;
+	re(x);
+	if (x)
+		return new tnode(x);
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:235]:
```suggestion
	if (x) return new tnode(x);
```

File: content/6_Advanced/Treaps.mdx [hunk 5]
```diff
@@ -207,7 +254,268 @@ int main() {
...
+	} else {
+		merge(right->left, left, right->left);
+		treap = right;
+	}
+	treap->size = 1 + size(treap->left) + size(treap->right);
+}
+
+ostream& operator<<(ostream &os, Node *n) {
+	if (!n)
+		return os;
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:485]:
```suggestion
	if (!n) return os;
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
> > at least your code is consistently bad
> 
> kinda rude

lmao

File: content/6_Advanced/Treaps.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx:201]: bruh what why

**Comment by dutinmeow** [comment]
> > at least your code is consistently bad
> 
> kinda rude

dw, I'm used to it 😞

File: content/6_Advanced/Treaps.mdx [hunk 6]
```diff
@@ -197,8 +197,7 @@ pt calc(pt x) {
 }
 	
 pair<pt, pt> split(pt t, int v) { // >= v goes to the right
-	if (!t)
-		return {t,t};
+	if (!t) return {t,t};
```
Comment (SansPapyrus683) [inline content/6_Advanced/Treaps.mdx:200]: bruh

Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx:200]: ~~add braces~~

Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:200]: don't u dare

Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx:200]: at least put a space after the comma like a normal person

Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:200]: but this isn't even my code...

Comment (Aadit-Ambadkar) [reply content/6_Advanced/Treaps.mdx]: you can still fix it

File: content/6_Advanced/Treaps.mdx [hunk 7]
```diff
@@ -207,7 +252,267 @@ int main() {
...
+	return os;
+}
+
+int main() {
+	int N, Q;
+	string S;
+
+	cin >> N >> Q >> S;
+	for (char c : S)
+		merge(root, root, new Node(c));
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:495]: please braces or one line

File: content/6_Advanced/Treaps.mdx [hunk 8]
```diff
@@ -147,50 +178,64 @@ ll ans, inv;
...
+		auto p = split(t->c[1], v);
+		t->c[1] = p.f;
+		return {calc(t), p.s};
 	}
 }
 
 pt merge(pt a, pt b) {
-	if (!a || !b) return a?:b;
+	if (!a || !b)
+		return a ? NULL : b;
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:214]: no revert this change. braces or one line

File: content/6_Advanced/Treaps.mdx [hunk 9]
```diff
@@ -147,50 +178,64 @@ ll ans, inv;
...
+	tnode (int val) {
+		pri = rng();
+		val = val;
+		sz = 1;
+		c[0] = c[1] = NULL;
 	}
 };
-int getsz(pt x) { return x?x->sz:0; }
+
+int getsz(pt x) { return x ? x->sz : 0; }
```
Comment (Aadit-Ambadkar) [inline content/6_Advanced/Treaps.mdx:192]: expand this

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Treaps.mdx (hunk 9 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:192]: What happened to braces or one line?

File: content/6_Advanced/Treaps.mdx (hunk 7 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx]: done

File: content/6_Advanced/Treaps.mdx (hunk 8 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx]: This isn't my code, you can go in and make any changes necessary

File: content/6_Advanced/Treaps.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx]: wait what why

File: content/6_Advanced/Treaps.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx]: maybe because functions can't not have braces?

Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx]: just change the code, I didn't even make this

File: content/6_Advanced/Treaps.mdx [hunk 10]
```diff
@@ -207,7 +255,267 @@ int main() {
...
+node to be its key. Because maintaining this value explicitely will result in up
+to $\mathcal{O}(n)$ values to be updated per insertion/deletion, we will keep
+this value implicitly.
+
+The index of a node is equal to the number of nodes less than it. It is
+important to note that these nodes can occur both in the left subtree of the
+current node as well as the node's ancestors and the left subtree of its
+ancestors. In other words,
+
+$$\texttt{index}(u) = |\{1 + \texttt{size}(P.\texttt{left}) \mid P \text{ is on path from } u \text{ to root}\}|$$
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:283]: Use

```
$$
equation
$$
```

to center an equation. Also, this isn't quite correct because you need to consider its children too.

File: content/6_Advanced/Treaps.mdx [hunk 11]
```diff
@@ -207,7 +255,267 @@ int main() {
...
+
+```cpp
+struct Node {
+	int val;
+	int weight, size;
+	Node *left, *right;
+	Node(int c) : val(c), weight(rand()), size(1), left(NULL), right(NULL) {}
+} *root;
+
+inline int size(Node *treap) {
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:362]: adding `inline` likely does nothing

File: content/6_Advanced/Treaps.mdx [hunk 12]
```diff
@@ -207,7 +255,267 @@ int main() {
...
+</LanguageSection>
+
+With split and merge, how do we implement the additional operations?
+
+- **Insert** can be done with one split and two merges: We split the array
+  between the index we want to insert, create a new node with the corresponding
+  value, and merge the three sections together.
+- **Delete** can be done with two splits and one merge: We split the array into
+  three parts before and after the index, and merge the two parts together.
+- **Range Queries** can be performed by maintaining additional data in each
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:414]:
```suggestion
- **Range queries** can be performed by maintaining additional data in each
```

File: content/6_Advanced/Treaps.mdx [hunk 13]
```diff
@@ -207,7 +255,267 @@ int main() {
...
+  node. We update this data whenever we update $\texttt{size}$.
+- **Range Updates** can be performed by maintaining a lazy tag in each node (as
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:416]:
```suggestion
- **Range updates** can be performed by maintaining a lazy tag in each node (as
```

File: content/6_Advanced/Treaps.mdx [hunk 14]
```diff
@@ -49,31 +49,62 @@ frequency: 2
 
 # Treaps
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:50]:
```suggestion
## Treaps
```

Comment (bqi343) [reply content/6_Advanced/Treaps.mdx:50]: same for the other headers

File: content/6_Advanced/Treaps.mdx [hunk 15]
```diff
@@ -49,31 +49,62 @@ frequency: 2
...
+$x \in \texttt{left}$ are strictly smaller than all keys $y \in \texttt{right}$.
+Furthermore, we need to merge these two treaps such that the resultant treap
+still satisfies the max heap property.
+
+We root the resultant treap at the root node that has the higher priority, and
+recursively call `merge` on the other tree and the corresponding subtree of the
+chosen tree.
 
-We root the resultant treap at the root node that has the higher priority, and recursively call `merge` on the other tree and the corresponding subtree of the chosen tree. 
 ## Implementation
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:108]: it would be great if this implementation actually solved the focus problem. Probably best to move the current focus problem somewhere else and find something new in place of it?

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by jessechoe10 — APPROVED** [review]

File: content/6_Advanced/Treaps.mdx (hunk 15 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx]: I'll move the code to a solution page, not sure what to replace it with tho. There aren't that many non-implicit treap problems

Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx]: Not sure why the commits are not syncing up, I've added the solution in a seperate PR https://github.com/cpinitiative/usaco-guide/pull/2258

**Review by Aadit-Ambadkar — APPROVED** [review]

File: content/6_Advanced/Treaps.mdx (hunk 15 above)
Comment (bqi343) [reply content/6_Advanced/Treaps.mdx]: wdym by the commits aren't syncing up?

**Review by bqi343 — APPROVED** [review]
